### PR TITLE
Improve large SDK thread performance

### PR DESCRIPTION
**Pull request by lsm**
Reduces large SDK thread CPU/UI churn by fixing both client render invalidation and daemon LiveQuery amplification.

Adds memoized message rendering, linear message-map construction, no-op delta suppression, SDK message window/expression indexes, batched hot-path sdk_messages writes, and direct UUID/status replay lookups to avoid repeated queued-message scans.

Tests: bun test packages/daemon/tests/unit/2-handlers/rpc-handlers/live-query-messages.test.ts packages/daemon/tests/unit/1-core/agent/sdk-message-handler.test.ts packages/daemon/tests/unit/4-space-storage/storage/reactive-database.test.ts; bun run check

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## Code Review — PR #1746: Improve large SDK thread performance

**Verdict: REQUEST CHANGES** — one P2 gap in the core client-side optimisation.

---

### What looks good

- **DB index** (`idx_sdk_messages_session_timestamp_id`): Genuine win — covers `session_id, timestamp DESC, id DESC` to satisfy the `top_level` CTE's `ORDER BY timestamp DESC, id DESC LIMIT ?2` without a separate sort pass. Migration 113 is correct and idempotent.
- **SQL IN → EXISTS**: Switches the subagent filter from `IN (SELECT id FROM tool_use_ids)` to a correlated `EXISTS`. Better for the query planner when `tool_use_ids` is large; correct result in all cases.
- **`useMessageMaps` O(n) fix**: The old double-loop was genuinely O(n²) on every render. The single-pass version with `lastUserUuid` + `pendingLeadingInits` preserves the semantics (verified by running the test suite and reviewing each edge case: leading inits, multiple inits between users, interleaved order).
- **`_applyMessagesDelta` no-op guard**: Only writing `this.sdkMessages.value = next` when `changed` is true prevents spurious Preact signal fires from LiveQuery events that touch no rows in the current window. Correct.
- **`prependMessages` dedup**: Prevents double-display of messages when paginated prepend overlaps with the LiveQuery window. Logic is sound.
- **`handleQuestionResolved` `useCallback`**: Pulls the inline callback out of the JSX and stabilises its reference; only re-created when `pendingQuestion` changes.

---

### P2 — `SDKMessageRenderer` memo is bypassed for the streaming hot path

`_applyMessagesDelta` always does:
```ts
let next = this.sdkMessages.value.slice();  // new array reference
// …
if (changed) { this.sdkMessages.value = next; }
```

So every LiveQuery delta that actually changes messages writes a **new array reference** to `sdkMessages`. `ChatContainer` reads that signal and passes it directly as `allMessages={messages}` to every `SDKMessageRenderer` in the list.

`areMessageRendererPropsEqual` compares:
```ts
prev.allMessages === next.allMessages
```

Because `allMessages` is a new array object on every delta, this check always returns `false` during streaming. **Every `SDKMessageRenderer` instance re-renders on every incoming message event** — which is exactly the O(N²) path driving 100% CPU on large threads.

The memo only helps for state changes that don't touch `messages` (question resolved, rewind-mode toggle, etc.). Those are rare. Streaming is not.

**Root cause**: `allMessages` is passed to `SDKAssistantMessage` solely to compute `hasSubagentChild` (one `.some()` scan). It isn't used at all in `SDKUserMessage` (destructured as `_allMessages` then discarded). Passing the entire array as a prop makes the memo always stale.

**Suggested fix** (minimal surface change):

1. In `ChatContainer`, compute a memoised `Set<string>` of assistant-message UUIDs that have at least one subagent child:
   ```ts
   const subagentParentIds = useMemo(() => {
     const s = new Set<string>();
     for (const m of messages) {
       const p = (m as SDKMessage & { parent_tool_use_id?: string }).parent_tool_use_id;
       if (p) s.add(p);
     }
     return s;
   }, [messages]);
   ```
   This `Set`'s **reference** is new on every render, but it contains the same values when nothing sub-agent-related changes. To make it stable, compare contents (or derive it from the existing `subagentMessagesMap` which is already memoised — its `.keys()` never changes unless the map itself changes).

2. Pass `subagentParentIds` (or derive it from `subagentMessagesMap`) to `SDKMessageRenderer` instead of `allMessages`.

3. Remove `allMessages` from `SDKMessageRenderer`'s `Props` and `areMessageRendererPropsEqual`.

With `allMessages` removed, the only props that change on a pure streaming append are `message` (for the newly added row) and `toolResultsMap`/`toolInputsMap`/`subagentMessagesMap` (only when tool blocks change). All existing message renderers except the new one skip re-rendering — reducing VDOM work from O(N²) to O(N) total.

---

### Minor observations (non-blocking)

- **Index redundancy**: `idx_sdk_messages_session` on `(session_id, timestamp)` already existed. The new index `idx_sdk_messages_session_timestamp_id` on `(session_id, timestamp DESC, id DESC)` adds the `id` column needed for stable pagination tie-breaking. Worth keeping, but a comment in `createIndexes` explaining why both exist would help future readers.
- **`pendingLeadingInits` test coverage**: The new performance test validates map sizes but doesn't explicitly cover the "init before first user" path. The existing test suite covers it implicitly, but an explicit case for `[init, user, init, user]` ordering would make the refactor's intent clear.
- **`SDKUserMessage` unused `_allMessages`**: `allMessages` is destructured but never read in `SDKUserMessage`. Once the P2 fix lands, this prop can be removed from `SDKUserMessage.Props` entirely.

File: packages/web/src/components/sdk/SDKMessageRenderer.tsx [hunk 1]
```diff
@@ -324,3 +325,35 @@ export function SDKMessageRenderer({
...
+		prev.sessionId === next.sessionId &&
+		prev.resolvedQuestions === next.resolvedQuestions &&
+		prev.pendingQuestion === next.pendingQuestion &&
+		prev.onQuestionResolved === next.onQuestionResolved &&
+		prev.onRewind === next.onRewind &&
+		prev.rewindingMessageUuid === next.rewindingMessageUuid &&
+		prev.rewindMode === next.rewindMode &&
+		prev.selectedMessages === next.selectedMessages &&
+		prev.onMessageCheckboxChange === next.onMessageCheckboxChange &&
+		prev.allMessages === next.allMessages &&
```
Comment (lsm) [inline packages/web/src/components/sdk/SDKMessageRenderer.tsx]: **P2 — Memo bypassed during streaming**

`allMessages` is a new array reference on every LiveQuery delta (since `_applyMessagesDelta` starts with `let next = this.sdkMessages.value.slice()`). So this reference-equality check returns `false` on **every** streaming event, and all N message renderers re-render on all N incoming events — preserving the O(N²) CPU path.

`allMessages` is only used inside `SDKAssistantMessage` to compute one boolean:

```ts
const hasSubagentChild = _allMessages?.some((msg) => {
  const p = (msg as SDKMessage & { parent_tool_use_id?: string }).parent_tool_use_id;
  return p === message.uuid;
}) || false;
```

**Suggested fix**: derive `hasSubagentChild` from the already-memoised `subagentMessagesMap` (which is keyed by `parent_tool_use_id`), remove `allMessages` from `SDKMessageRenderer`'s Props, and drop this comparator line. With `allMessages` gone, the only prop that changes on a pure streaming append is the new message object itself — all existing renderers skip.

Comment (lsm) [reply packages/web/src/components/sdk/SDKMessageRenderer.tsx]: Fixed in 651a8dbef. SDKAssistantMessage now derives hasSubagentChild from the existing subagentMessagesMap, and the allMessages prop was removed from ChatContainer, SDKMessageRenderer, SDKAssistantMessage, SDKUserMessage, and the memo comparator. This keeps existing message rows eligible for the memo fast path during streaming appends.

**Review by lsm — COMMENTED** [review]
## Approval — PR #1746

P2 finding fully addressed. All checks pass.

**What the fix does**: `allMessages` is removed from the renderer chain entirely — dropped from `SDKMessageRenderer.Props`, `SDKAssistantMessage.Props`, `SDKUserMessage.Props`, and `ChatContainer`'s JSX. `SDKAssistantMessage.hasSubagentChild` now derives its value from `subagentMessagesMap.has(message.uuid)`, which is semantically identical to the old `.some()` scan (the map is keyed by `parent_tool_use_id`) and uses the already-memoised map instead of the whole messages array.

The `areMessageRendererPropsEqual` comparator no longer includes `allMessages`, so a pure streaming append — which only changes the `message` prop for the newly added row and the shared map refs when content actually changes — leaves all existing renderers' props equal and they skip re-rendering.

Everything in the first round's approval list still holds. Merged net of changes:
- DB covering index for `(session_id, timestamp DESC, id DESC)` — avoids sort pass on the `top_level` CTE window query.
- SQL `IN → EXISTS` for the subagent filter.
- `sessionInfoMap` O(n²) → O(n).
- `_applyMessagesDelta` no-op guard — only fires the Preact signal when rows actually change.
- `prependMessages` dedup — prevents double-display on paginated prepend.
- `handleQuestionResolved` `useCallback` — stable callback reference.
- `SDKMessageRenderer` memo — now effective for the streaming hot path.

LGTM ✅

**Comment by lsm** [comment]
## Daemon CPU Investigation — Follow-up Findings

PR #1746's server-side changes (index + IN→EXISTS) are **marginal for daemon CPU**. The real bottleneck is LiveQuery re-evaluation amplification. Full investigation below.

### Core Finding: ~31 LiveQuery re-evaluations per agent turn

Every write to `sdk_messages` triggers a **full re-execution** of the `messages.bySession` CTE query. During a turn with 10 tool calls:

```
saveSDKMessage (tool_use)     → LiveQuery re-eval #1
updateMessageStatus           → LiveQuery re-eval #2  
updateMessageTimestamp        → LiveQuery re-eval #3
saveSDKMessage (tool_result)  → LiveQuery re-eval #4
updateMessageStatus           → LiveQuery re-eval #5
updateMessageTimestamp        → LiveQuery re-eval #6
... × 10 tool calls = ~31 full query re-evaluations per turn
```

Each re-evaluation does:
1. Two full scans of `sdk_messages` with `json_extract()` on every row
2. `JSON.stringify(all 200 rows)` for hash computation (~2MB, 5-20ms)
3. `computeDiff()` with per-row `JSON.stringify` comparison

**Estimated ~620ms of pure daemon CPU per turn** for a 500-message session.

The `beginTransaction`/`commitTransaction` batching mechanism already exists in `ReactiveDatabase` but is **never used** on the message-insertion hot path.

### Action Items

| Priority | Fix | Impact |
|----------|-----|--------|
| **P0** | Wrap message saves in `beginTransaction`/`commitTransaction` during a turn — collapses ~31 re-evaluations to 1 | ~90% reduction in LiveQuery CPU |
| **P0** | Expression index: `ON sdk_messages(session_id, json_extract(sdk_message, '$.parent_tool_use_id'))` | Eliminates per-row JSON parsing in both CTEs |
| **P1** | Replace `hashRows()` full-result-set serialization with row-level tracking | Eliminates 2MB+ stringify per re-eval |
| **P1** | Consolidate dual delta path (LiveQuery + MessageHub `state.sdkMessages.delta`) | Halves WebSocket payload volume |
| **P2** | Fix `acknowledgePersistedUserMessage` triple `getMessagesByStatus` + `.find()` | Eliminates O(n) JSON.parse per user message |
| **P2** | WebSocket `permessage-deflate` compression | 5-10x wire size reduction |
| **P3** | Debounce `broadcastSessionStateChange` for rapid-fire `context.updated` events | Reduces redundant state serialization |

Source files investigated: `reactive-database.ts`, `live-query.ts`, `live-query-handlers.ts`, `sdk-message-handler.ts`, `sdk-message-repository.ts`, `state-manager.ts`, `app.ts`.

File: packages/daemon/src/lib/agent/sdk-message-handler.ts [hunk 2]
```diff
@@ -314,8 +314,16 @@ export class SDKMessageHandler {
 			.getMessagesByStatus(session.id, 'enqueued')
 			.filter((enqueued) => isSDKUserMessage(enqueued));
 
+		if (enqueuedUsers.length > 0) {
+			this.withDbChangeBatch(() => {
+				db.updateMessageStatus(
+					enqueuedUsers.map((enqueuedUser) => enqueuedUser.dbId),
+					'consumed'
+				);
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/agent/sdk-message-handler.ts]: Fixed. I reverted the turn-end fallback path to update each enqueued message's status immediately before its own `messages.statusChanged`/`state.sdkMessages.delta` publish, so a publish failure can no longer pre-consume later queued messages. The hot replay/yield paths still keep their per-message status+timestamp pairs batched.

**Review by lsm — COMMENTED** [review]
## Approval — PR #1746 (Round 3)

All daemon-side fixes verified. All tests pass (106 total across 4 test files), full check suite clean.

### What's new since round 2

**P0 — Transaction batching (`withDbChangeBatch`)**
Wraps `saveSDKMessage`, `updateMessageStatus` + `updateMessageTimestamp` pairs in `beginTransaction`/`commitTransaction` on the hot path. The `ReactiveDatabase` proxy correctly intercepts `beginTransaction`/`commitTransaction`/`abortTransaction` (lines 125-127) before the general `METHOD_TABLE_MAP` lookup, so writes inside a batch accumulate `pendingTables` and flush as a single change event on commit. This collapses ~31 LiveQuery re-evaluations per turn to 1.

Correctly NOT applied to `acknowledgeOldestQueuedUserOnTurnEnd` — per-message atomicity is preserved there per review feedback.

**P0 — Expression indexes**
- `idx_sdk_messages_parent_tool` on `(session_id, json_extract(sdk_message, '$.parent_tool_use_id'))` — eliminates per-row JSON parsing in both the `top_level` and `subagent` CTEs.
- `idx_sdk_messages_uuid_status` on `(session_id, send_status, json_extract(sdk_message, '$.uuid'))` — covers the new direct UUID lookup.

Both verified by the `EXPLAIN QUERY PLAN` test asserting index usage.

**P2 — Direct UUID/status lookup (`getMessageByStatusAndUuid`)**
Replaces the triple `getMessagesByStatus` + `.find()` pattern with a single `WHERE session_id = ? AND send_status = ? AND json_extract(sdk_message, '$.uuid') = ?` query. Also extracted `inflatePersistedMessage` helper to DRY up the row mapping. All 4 callers in `sdk-message-handler.ts` updated.

### Minor note (non-blocking)
`acknowledgeOldestQueuedUserOnTurnEnd` still uses `getMessagesByStatus` (bulk scan) — acceptable since it's a fallback path that runs at most once per turn and typically processes 0-1 messages. Not worth optimizing further.

LGTM ✅
